### ISSUE TITLE
Inference failing for conditional types in function parameters

### ISSUE THREAD

**Issue by squidfunk** [microsoft/TypeScript#33369]
When using generics with conditionals in a function's signature, type inference seems to break down. I narrowed the problem down to the following reproducible case. Interestingly, the issues vanishes when another generic "helper" parameter is introduced, as can be seen in the example.

**TypeScript Version:**  3.6.3, 3.6.2, 3.5.x

**Search Terms:** parameter, inference, functions, generic, conditional types

**Code**

```ts
export interface Foo<T> {
  bar: T
}

export type FooType<T> = T extends Foo<infer U> ? U : never
export type FooLike<T> = T extends Foo<FooType<T>> ? T : never

export type FooQueryType<T> =
  T extends (foo: infer U, ...args: any[]) => boolean
    ? U extends FooLike<infer V>
      ? V
      : never
    : never

export type FooQueryParameters<T> =
  T extends (foo: any, ...args: infer U) => boolean
    ? U
    : never

export type FooQueryLike<T> =
  T extends (foo: FooQueryType<T>, ...args: any[]) => boolean
    ? T
    : never

export type FooQuery<T> = (foo: FooLike<T>) => boolean

export function matchWithBrokenInference<T>(
  fn: FooQueryLike<T>, ...args: FooQueryParameters<T>
): FooQuery<FooQueryType<T>> {
  return foo => fn(foo, ...args)
}

export function matchWithCorrectInference<T, U extends FooQueryLike<T>>(
  fn: T, ...args: FooQueryParameters<U>
): FooQuery<FooQueryType<U>> {
  return foo => (fn as U)(foo, ...args)
}

function query<T extends Foo<string>>(foo: FooLike<T>, data: string) {
  return true
}

const q1 = matchWithBrokenInference(query, "test") // incorrect inference of query
const q2 = matchWithCorrectInference(query, "test")
```

**Expected behavior:**

`matchWithBrokenInference` infers `query` correctly.

**Actual behavior:**

`matchWithBrokenInference` infers `query` incorrectly.

**Playground Link:** [reproducible example](http://www.typescriptlang.org/play/#code/KYDwDg9gTgLgBASwHY2FAZgQwMbDgMQggB4AVAPjgG8AoOOAI0ygC45SaBfGm0SWODACeYPIQikRwMpQC87OKFRIAJgGcCRYsnRo4AVUoB+A3DZJgANzS9w0eMNGaIAGQQBraRTjzSikMrqzsTikqIyxgrmVja2-A5SzgCKAK5oQmFecnQKSsCqGgAU6ERsOnr6ADRwAHR1zADmamyYSEIA2gC6AJQ+lAxEADbArTn0Jvr+gRribp7aSLpQcABq5GP0cCYrG-TR1lAb+7F89oKJ4qnpAArMmAC2wKhQajI+OX55BXDFpXCtQmqdRqjWaiEWFV6sn6QxGSA2EyOcAsBx4pwEjjERCuUCEcyy73onwC+SCvwgbEuaVxmRkQPqUCaLTaXShMIgw1Gmy27CRKJOdgxF2x1KEb3k5MpRHxMjZjFhoziZ3QKSQ2BgCAgSDg90wMGwAAsAOoIGAGgBCUAgniQAEkIVB8rgZIUcugkFKIDi8R4svSQYywVSbndHs9XhQaN1Pd6QiL0rSKJRaPRHTAUlBtSUIH04O7yf7Qd0uGjBfAVWqNVqdXrDSazQBhaCO9X2pZOrzVSZfILB3EypOu+jutikQuBmOi25QB5PNCvQxRyfpONe0W0wzJnJpjNZoi54razAafTdAu1BlNYvcGgV9Wa7UAR1FZCmpJmWjUMCgyAa5HIkrOAO5DVCoeqYGwX4-kgDS9CmcA7pmghQGkJY0NgWpfnAj4AIw+DW+rGqaFpWjabZoB2hTPuk1QAESoF+tG9AA9Mx4IYVALbwOUjpqngEDoNhoroZh8CPgATPhuqEfWBpNpxwCtg6lHUbidEMTATE0EAA)

**Related Issues:** None found

**Comment by AnyhowStep** [comment]
This type,
```ts
type FooQueryLike<T> =
  T extends (foo: FooQueryType<T>, ...args: any[]) => boolean
    ? T
    : never
```

And your usage of it are very suspicious.

**Comment by AnyhowStep** [comment]
Okay, I tried to simplify on mobile, lol,

```ts
export interface Foo<T=any> {
  bar: T
}

type Query<T=any> = (foo: Foo<T>, ...args: any[]) => boolean;

export type QueryType<T extends Query> =
  T extends (foo: Foo<infer U>, ...args: any[]) => boolean ?
    U :
    never

export type QueryParameters<T extends Query> =
  T extends (foo: any, ...args: infer U) => boolean
    ? U
    : never

export type Result<T extends Foo> = (foo: T) => boolean

declare function match<T extends Query> (query : T, ...args : QueryParameters<T>) : (
  Result<QueryType<T>>
);

```

**Comment by AnyhowStep** [comment]
I'm even tempted to say this should work,

```ts
export interface Foo<T=any> {
  bar: T
}

export type Result<T extends Foo> = (foo: T) => boolean

declare function match<
  FooT extends Foo,
  ParamsT extends readonly any[]
> (
  query : (foo : FooT, ...args : ParamsT) => boolean, 
  ...args : ParamsT
) : (
  Result<FooT>
);
```

However, I'm on mobile and can't test that

**Comment by AnyhowStep** [comment]
I just tested and the type of `foo` is preserved,
```ts
export interface Foo<T=any> {
  bar: T
}

export type Result<FooT extends Foo> = (foo: FooT) => boolean

declare function match<
  FooT extends Foo,
  ParamsT extends readonly any[]
> (
  query : (foo : FooT, ...args : ParamsT) => boolean, 
  ...args : ParamsT
) : (
  Result<FooT>
);

interface MyFooLike {
    bar : "haha",
    baz : "hehe",
}
declare function myQuery (
    foo : MyFooLike,
    arg0 : number,
    arg1 : Date
) : boolean;

//const result: Result<MyFooLike>
const result = match(myQuery, 3.141, new Date());
//type resultParams = [MyFooLike]
type resultParams = Parameters<typeof result>;
```
[Playground](http://www.typescriptlang.org/play/#code/KYDwDg9gTgLgBASwHY2FAZgQwMbDgMQggB4AVAXkyQE8A+OAbwCg44AjTKALjlKYF8mTUJFhwY1MHgBKwAM4BXADYxihCKTihUSACZyCReuTgAKdER7rSASjjl6bIkuBUhu4NiWc86BUmwYBAgkOABbTBhsAAtiFkMNLRAdfQSAGniABU5MMLlNbWA9AyhXXRClajgqagBtAF0melN4gEcFNCqecyI4KyJSNLgAOlHOAHMDHmyoXPy7B3ZnVyQh+NHhiam4Gbm+O2742UUVNQHaJhsAbiFkVAwcPABZanUAGQQAazxmVlYOKB9OAAImimDBwIyf3YmAAXkDQcBosBIQImB4vD44H4AkEQuFqABFDpQKotaEWCBAl7vL7AKF-CYABiBSAUYTYaAZrAmAEYgQARSLAS5ApwQFxUG5MAD0MuwITk8FKJxgPGOylUNKIH2+FwVSCVcBVmvs4UiMVMYSJJOoQwAzMNeQAWXlDJDAADucCFqFMNmusplEikxvkmt2eTNtW1EF1wEaIbwJpUkYMJkjwHucmISYg6DDqtoVyAA)

**Comment by weswigham** [comment]
OK, so the issue is due to _contextual typing_; specifically recent additions around generic signatures. The contextual type of `query` in the functioning case is `T` - `T` is unconstrained and has no signatures, which means we forward along the type of `query` unmodified into inference (which is then immediately matched against `T`). In the "broken" case, the argument type is `FooQueryLike<T>`, which _does_ have signatures, which causes us to go "ah, we may need to unify the type parameters in `query` with the arguments of `FooQueryLike`. Let's defer inference to them until the next inference stage". We then produce `unknown` as the inference result for `T` during that first stage, since we made no successful inferences. Before moving onto the next stage, we check that none of our inferences have already made the signature invalid, but, whatdoyouknow, that `unknown` (from having made no inferences whatsoever), when fed through the arguments, becomes `never` for the first argument and triggers an error, causing us to bail on the second inference phase and immediately issue an error. _Generally speaking_ we assume that when a type parameter is instantiated with it's constraint, a call should succeed, as the constraint it supposedly strictly "broader" than the type parameter itself; but conditionals throw a massive wrench into this, as the constraint of an unconstrained type parameter isn't going to `extend` anything in a conditional, which means those conditions are always going to evaluate to their `false` branches, which, more often than not, will produce a `never` which in turn disallows any actual assignments.

**Comment by weswigham** [comment]
Additionally, even if we got past that, when we go to compare `<QueryT extends Foo<string>>(foo: FooLike<QueryT>, data: string) => boolean`, because of the contextual signature, we erase it down to `(foo: FooLike<QueryT>, data: string) => boolean` and compare against `FooQueryLike<QueryT>` (hoisting the type param to the outside so we can try to move it to the return type); however `(foo: FooLike<QueryT>, data: string) => boolean` ends up _not_ being assignable to `FooQueryLike<QueryT>` _because_ `FooQueryLike<QueryT>` is a generic conditional and nothing is assignable to it (this is an outstanding bug)! The case where we _don't_ rip off the type parameter and leave it in place causes us to convert the signature to its' "canonical signature", which involves mapping all the local type params to constraints (and instantiating the target in the context of that type); consequently, that _removes_ the generics from the equation, `FooQueryLike` is evaluated to a concrete type, and the assignment checks out.

**Comment by weswigham** [comment]
While a fix to #26933 won't also fix this, it is a prerequisite, so this is blocked on #26933.

**Comment by RyanCavanaugh** [comment]
Based on the above explanation, I'm going to track this as a duplicate of #26933
